### ISSUE TITLE
[STON] Ston materialization fails for sorted collections with references to other graph objects

### ISSUE THREAD

**Issue by pavel-krivanek** [guillep/pharo-core#9390]
**Bug description**
If in an object graph, there is an sorted collection that points to other objects in the object graph. The STON materialization fails because it tries to work with STON references as with materialized objects.

**To Reproduce**
1. file-in this simple example: [ObjectWithSortedCollection.zip](https://github.com/pharo-project/pharo/files/6590190/ObjectWithSortedCollection.zip).
2. execute  `ObjectWithSortedCollection issue`
3. `Instance of STONReference did not understand #'<='`

Basically, it does this: 
```smalltalk
	data := (1 to: 2) collect: [ :e | ObjectWithSortedCollection new ].

	data do: [ :each | 
		data do: [ :object | 
			each sortedCollection add: object ] ].

	serialized := STON toString: data.
	materialized := STON fromString: serialized.
```

![image](https://user-images.githubusercontent.com/572494/120616376-76d90680-c459-11eb-83cf-af26bb8d2b18.png)

**Comment by pavel-krivanek** [comment]
@svenvc

**Comment by svenvc** [comment]
Hi,

I tried the following, which seems to work (Pharo 9, latest STON from https://github.com/svenvc/ston

```
data := (1 to: 2) collect: [ :e | e -> SortedCollection new ].

data do: [ :each | 
	data do: [ :object | 
		each value add: object ] ].

serialized := STON toString: data.
materialized := STON fromString: serialized.
```

In any case, #add: or #addAll: should not be called on SortedCollection at reconstruction time as it results in resorting.

I can't remember, but maybe the extensions on SortedCollection are missing in a stock Pharo image.

Sven


<blockquote><img src="https://opengraph.githubassets.com/59df2f13b862c6b1a499ddada77ad08409b4ad803c3144411a587d192b3a2e98/svenvc/ston" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/svenvc/ston">svenvc/ston</a></strong></div><div>STON - Smalltalk Object Notation - A lightweight text-based, human-readable data interchange format for class-based object-oriented languages like Smalltalk. - svenvc/ston</div></blockquote>

**Comment by svenvc** [comment]
It is indeed a recent addition:

https://github.com/svenvc/ston/commit/3565d388172b76c180454575d4c2f71019f130c4

**Comment by Ducasse** [comment]
So sven should we integrate a more recent version for Pharo 90?

**Comment by pavel-krivanek** [comment]
Ah, well... then Ston needs to be updated in Pharo... (in plain Pharo 9 it fails)

(I know that this way of filling of sorted collection is stupid, I did it just for demonstration purpuses)

**Comment by svenvc** [comment]
@Ducasse All my libraries run fine on Pharo 7, 8 and 9 with CI tests all green. In principle they can just be loaded.

The problem is, they might overwrite some change in Pharo 9 itself that was not applied to the upstream repository. Normally I try to track those, but sometimes I miss them. It requires a full manual check, there is some way to do this with Iceberg, but I more or less forgot how to do this.

**Comment by Ducasse** [comment]
I know. @estebanlm is working on a process to avoid that people commit in Pharo libraries externally developed. 
We want to generalise it from Spec2, Iceberg to all the components.

**Comment by estebanlm** [comment]
oh, mechanism is already there, but I am trying it before generalize it or propose it to everybody.
take a look at the document attached here, for an explanation: http://forum.world.st/Pharo-Progress-Report-2021-06-02-td5130217.html


<blockquote><div><strong><a href="http://forum.world.st/Pharo-Progress-Report-2021-06-02-td5130217.html">Pharo Smalltalk Developers - Pharo Progress Report 2021-06-02</a></strong></div><div>Pharo Progress Report 2021-06-02. Hi, We finished one short term task : Improve integration structure of Spec2 and NewTools. I am adding an explanation of the change at the end of this report. The...</div></blockquote>

**Comment by MarcusDenker** [comment]
Do we need to do something for Pharo9?

**Comment by svenvc** [comment]
It would be best to merge the latest STON into Pharo 9.

This requires a manual check to see if there are changes to STON in Pharo 9 that were not applied to STON upstream, else they would get lost.

I forgot how to do this efficiently.

**Comment by svenvc** [comment]
Thank you all !
